src/QsCompiler/Transformations/ContentLifting.cs

```diff
@@ -0,0 +1,547 @@
+                        ExpressionKind.UnitValue,
+                        TypeArgsResolution.Empty,
+                        ResolvedType.New(ResolvedTypeKind.UnitType),
+                        new InferredExpressionInformation(false, false),
+                        QsNullable<Tuple<QsPositionInfo, QsPositionInfo>>.Null);
+                }
+
+                var call = new TypedExpression(
+                    ExpressionKind.NewCallLikeExpression(generatedOpId, arguments),
+                    typeArguments.IsNull
+                        ? TypeArgsResolution.Empty
+                        : typeArguments.Item
+                            .Select(type => Tuple.Create(CurrentCallable.Callable.FullName, ((ResolvedTypeKind.TypeParameter)type.Resolution).Item.TypeName, type))
+                            .ToImmutableArray(),
+                    ResolvedType.New(ResolvedTypeKind.UnitType),
+                    new InferredExpressionInformation(false, true),
+                    QsNullable<Tuple<QsPositionInfo, QsPositionInfo>>.Null);
+
+                return (generatedOp, new QsStatement(
+                    QsStatementKind.NewQsExpressionStatement(call),
+                    LocalDeclarations.Empty,
+                    QsNullable<QsLocation>.Null,
+                    QsComments.Empty));
+            }
+        }
+
+        protected LiftContent() : base(new TransformationState())
+        {
+            this.Namespaces = new NamespaceTransformation(this);
+            this.StatementKinds = new StatementKindTransformation(this);
+            this.Expressions = new ExpressionTransformation(this);
+            this.ExpressionKinds = new ExpressionKindTransformation(this);
+            this.Types = new TypeTransformation<TransformationState>(this, TransformationOptions.Disabled);
+        }
+
+        protected class NamespaceTransformation : NamespaceTransformation<TransformationState>
+        {
+            public NamespaceTransformation(SyntaxTreeTransformation<TransformationState> parent) : base(parent) { }
+
+            public override QsCallable OnCallableDeclaration(QsCallable c)
+            {
+                SharedState.CurrentCallable = new CallableDetails(c);
+                return base.OnCallableDeclaration(c);
+            }
+
+            public override QsSpecialization OnBodySpecialization(QsSpecialization spec)
+            {
+                SharedState.InBody = true;
+                var rtrn = base.OnBodySpecialization(spec);
+                SharedState.InBody = false;
+                return rtrn;
+            }
+
+            public override QsSpecialization OnAdjointSpecialization(QsSpecialization spec)
+            {
+                SharedState.InAdjoint = true;
+                var rtrn = base.OnAdjointSpecialization(spec);
+                SharedState.InAdjoint = false;
+                return rtrn;
+            }
+
+            public override QsSpecialization OnControlledSpecialization(QsSpecialization spec)
```

Comment:
The original idea was for these flags that tell what specialization you are in were implementation details and supposed to be private to the class. When that was the case, there was no reason to make one for ControlledAdjoint because the logic didn't need it. But because of the nature of the TransformationState class being public, these flags are now exposed and can be used by other logic outside the class, or in inherited classes. If other classes can write logic around these flags, I guess its fine to have another flag for ConrolledAdjoint, even though it is not used in this transformation's logic.

I believe you added this in your PR.

Reply:
Yup, I did.